src/brushless_arm.cpp

```diff
@@ -1,17 +1,271 @@
-#ifdef BRUSHLESS_ARM //LEAVE THIS AT THE TOP OF THIS FILE
-
+#ifdef BRUSHLESS_ARM // LEAVE THIS AT THE TOP OF THIS FILE
+#include "rover_arm.h"
+// Drivers.
+#include "RoverArmMotor.h"
+#include "AMT22.h"
 #include <Arduino.h>
-#include "brushless_arm.h"
+#include <SPI.h>
+#include <EEPROM.h>
+
+// Standard includes.
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <bitset>
+#include <limits>
+
+float armBrushlessTargetAngles[3];
+float armBrushlessActualAngles[3];
+
+static unsigned long lastTime;
+
+#define MIN_FLOAT -std::numeric_limits<float>::infinity()
+#define MAX_FLOAT std::numeric_limits<float>::infinity()
+static void attach_all_interrupts();
+
+#define DEBOUNCE_DELAY 100
+
+volatile int limit_elbow_max_activated = 0;
+volatile int limit_elbow_min_activated = 0;
+volatile int limit_shoulder_max_activated = 0;
+volatile int limit_shoulder_min_activated = 0;
+volatile int limit_waist_max_activated = 0;
+volatile int limit_waist_min_activated = 0;
+
+volatile unsigned long last_trigger_time_elbow_max = 0;
+volatile unsigned long last_trigger_time_elbow_min = 0;
+volatile unsigned long last_trigger_time_shoulder_max = 0;
+volatile unsigned long last_trigger_time_shoulder_min = 0;
+volatile unsigned long last_trigger_time_waist_max = 0;
+volatile unsigned long last_trigger_time_waist_min = 0;
+
+/*---------------------ELBOW_SERVO DECLARATIONS---------------------*/
+#if TEST_ELBOW_SERVO == 1
+RoverArmMotor Elbow(&armBrushlessTargetAngles[0], &armBrushlessActualAngles[0], PWM2, -1, CS1, BLUE_ROBOTICS, ELBOW_MIN_ANGLE, ELBOW_MAX_ANGLE);
+#endif
+
+/*---------------------SHOULDER_SERVO DECLARATIONS---------------------*/
+#if TEST_SHOULDER_SERVO == 1
+RoverArmMotor Shoulder(&armBrushlessTargetAngles[1], &armBrushlessActualAngles[1], PWM1, -1, CS2, BLUE_ROBOTICS, SHOULDER_MIN_ANGLE, SHOULDER_MAX_ANGLE);
+#endif
+
+/*---------------------WAIST_SERVO DECLARATIONS---------------------*/
+#if TEST_WAIST_SERVO == 1
+RoverArmMotor Waist(&armBrushlessTargetAngles[2], &armBrushlessActualAngles[2], PWM3, -1, CS3, BLUE_ROBOTICS, WAIST_MIN_ANGLE, WAIST_MAX_ANGLE);
+#endif
+
+void brushless_arm_setup()
+{
+  // Initiate SPI bus.
+  SPI.begin();
+  SPI.setClockDivider(SPI_CLOCK_DIV64);
+
+  /* ELBOW_SERVO setup */
+#if TEST_ELBOW_SERVO == 1
+  Elbow.setAngleLimits(ELBOW_MIN_ANGLE, ELBOW_MAX_ANGLE);
+  Elbow.stop_tick = 1;
+  Elbow.fight_gravity = 1;
+  Waist.error_range = 3.0f;
```

Comment:
the encoders have a 0.2deg accuracy, we are able to close the window of "close enough" much more